src/detail/telemetry_scraper.cc

```diff
@@ -0,0 +1,132 @@
+#include "broker/detail/telemetry_scraper.hh"
+
+#include <caf/actor_system_config.hpp>
+#include <caf/behavior.hpp>
+#include <caf/event_based_actor.hpp>
+#include <caf/telemetry/metric.hpp>
+#include <caf/telemetry/metric_registry.hpp>
+
+#include "broker/defaults.hh"
+#include "broker/logger.hh"
+#include "broker/message.hh"
+
+namespace ct = caf::telemetry;
+
+namespace broker::detail {
+
+namespace {
+
+table to_table(const std::vector<ct::label>& labels) {
+  auto str = [](auto str_view) {
+    return std::string{str_view.begin(), str_view.end()};
+  };
+  table result;
+  for (const auto& lbl : labels)
+    result.emplace(str(lbl.name()), str(lbl.value()));
+  return result;
+}
+
+} // namespace
+
+// -- initialization -----------------------------------------------------------
+
+telemetry_scraper_state::telemetry_scraper_state(
+  caf::event_based_actor* self, caf::actor core,
+  std::vector<std::string> selected_prefixes, caf::timespan interval,
+  topic target)
+  : self(self),
+    core(std::move(core)),
+    selected_prefixes(std::move(selected_prefixes)),
+    interval(interval),
+    target(std::move(target)) {
+  // nop
+}
+
+telemetry_scraper_state::telemetry_scraper_state(caf::event_based_actor* self,
+                                                 caf::actor core, topic target)
+  : self(self), core(std::move(core)), target(std::move(target)) {
+  using svec = std::vector<std::string>;
+  const auto& cfg = self->system().config();
+  if (auto filter = caf::get_as<svec>(cfg, "broker.metrics-publish-filter"))
+    selected_prefixes = std::move(*filter);
+  interval = caf::get_or(cfg, "broker.metrics-publish-interval",
+                         defaults::metrics_publish_interval);
+}
+
+caf::behavior telemetry_scraper_state::make_behavior() {
+  if (!core) {
+    BROKER_WARNING("spawned a telemetry scraper with invalid core handle");
+    return {};
+  }
+  self->monitor(core);
+  self->set_down_handler([this](const caf::down_msg& down) {
+    if (down.source == core)
+      self->quit(down.reason);
+  });
+  if (interval.count() > 0) {
+    last_scrape = self->clock().now();
+    self->scheduled_send(self, last_scrape + interval, caf::tick_atom_v);
+  }
+  return {
+    [this](caf::tick_atom) {
+      scrape();
+      if (!tbl.empty())
+        self->send(core, atom::publish_v, make_data_message(target, tbl));
+      last_scrape += interval;
+      auto next_scrape = last_scrape + interval;
+      auto now = self->clock().now();
+      while (now >= next_scrape)
+        next_scrape += interval;
```

Comment:
Yeah, it does look a bit convoluted at a first glance. Why not just use `delayed_send`? So here's my reasoning to do it this way.

When using `delayed_send` (or manually scheduling based on `now` + some interval), you'll always have some drift over time because the messages never arrive at precisely the scheduled time and the time for processing the timeout itself is also subject to variations (unpredictable interrupts, shifting cache hit rates, and so on). Say our first scrape was at precisely 09:00:00 and your interval is 5s. With this code, Broker tries to scrape at ...:05, ...:10, and so on. If one scrape gets delayed (because of scheduling artifacts or load peaks), then at least the next one should be back on time.

A drift over time is no big deal on a single system. However, with multiple distributed workers and each drifting slightly different over time, I think it's worth the bit of extra bookkeeping to try and have each worker publish its data on a somewhat predictable schedule.